### PR TITLE
Create a new EoC function for checking if NPC follower is present

### LINKED ISSUE #73140: code for followers to chime in
### Is your feature request related to a problem? Please describe.

I really want to write humorous conversations between various NPC followers, or to have NPC followers chime in on conversations with other NPCs. However, currently, we have no reliable way to test if a specific follower is around to chime in. This has been sorely needed for years, but now that I'm working on making Liam super-detailed (#73133, #73108) it's getting more necessary

### Solution you would like.

Probably we need a pair of conditional variables: "NC_present" and "follower_present". Both of these look for NPC class IDs, because that's something stored in a way we can search, and is generally unique for most unique NPCs.

here's an example of how this might look, when you ask Smokes if he always greets newcomers with "sorry we're full".

```
  {
    "id": "TALK_FREE_MERCHANTS_MERCHANT_AboutGreeting",
    "type": "talk_topic",
    "dynamic_line": {
      "concatenate": [
       "A lot of the time, yes.  Our address is in the computers of most of the evac shelters around here, and people keep showing up that we have to turn away.  It gets easier when you have something you can read off a script.",
        {
          "NC_present": "NC_BEGGAR_3",
          "yes": "\"\n\nYou notice a few of the tired looking people in the room behind you glancing around awkwardly at the comment.  One of them scowls openly in your direction.  Smokes frowns.  \"Not that much easier, though."
        },
        {
          "follower_present": "NC_BEGGAR_3",
          "yes": "\"\n\nLuo sighs.  \"Can't believe you haven't just posted a sign outside by now.  Or that you still have to say it, even after all this time.",
          "no" : {
              "follower_present": "NC_STARTINGNPC_LIAM",
              "yes": "\"\n\nLiam frowns  \"That's rough, buddy. Classic customer service stuff, amirite.  Kinda thought we'd be done with that when the world ended."
          }
        },
        {
        }
      ]
    }
```
What this would do is search for an NPC with the class NC_BEGGAR_3, within line of sight and some fairly large distance like 10 squares, either following the player (follower_present) or not following the player (NC_present) - for the latter, we'd also check that they're not hostile to the player I think. The reason for "follower_present" specifically is because most use cases of this are going to want the NPC to be a follower, but we don't want this to trigger if they're not. For example, if you talk to Smokes while Luo is a beggar in the lobby, we don't necessarily want her to chime in with the same dialogue she would chime in with if she was your follower later in the game, as I've demonstrated in the example.

Also for fun, I expanded the example to show how the conversation might include options from more than one follower.

### Describe alternatives you have considered.

I considered something a little more bespoke than a dialogue conditional but after reviewing how it would work, I think this is actually a better set up than my other methods anyway.

### Additional context

As always the next big barrier would be to go through adding a whole bunch of this kind of context dialogue, so that it actually feels like your NPCs are present in the room while you're doing stuff.

### PR DESCRIPTION
**Pull request by katemonster33**
#### Summary
Features "Add new function follower_present for use in JSON"

#### Purpose of change
Fixes: #73140 

#### Describe the solution
Create a new function follower_present and add it to the list of condition EoC's for usage in JSON

##### Erk's additions
Erk here. To demo Kate's new additions I added some dialogue using them:

- Liam and Luo, if followers, can react to Smokes talking about the beggars in the main room. Liam shows some distaste for how they're being treated because he's just kind of a chill dude, but Smokes has no options here. Luo wants to be mad at you for taking money for helping them, but even she can't find it in herself to be grouchy on that one.
- When doing exodii mission 3 and asking for directions, various followers have a chance to notice good or bad directions. Liam can point out the right ones, and the merc, yusuke the beggar-scavenger, and chris isherwood can all spot bad ones. They're not guaranteed to do so but the more 'explorer' followers you've got the better chance you have of getting a tip-off and saving some effort. I'm really happy with this one, as it showcases some ideas for how having followers could have more impact on the game in meaningful ways. Other major NPCs could be added to the list. Theoretically we could even have some of the random NPCs kick in here but it would be harder to write since their names are random.
- Liam has a funny 'wingman' response to your, um, tryst with Alonso. I only added this because it makes me laugh. Your mileage may vary. I am not a classy man.

Erk out. Back to your regularly scheduled PR.

#### Testing
Create new game in "friends til the end" scenario, spawn the evac center nearby, drive Liam there, talk to the shop keep, see that Liam chimes in.

#### Additional context
in free_merchant_shopkeep_talk.json
```javascript
  {
    "id": "TALK_FREE_MERCHANTS_MERCHANT_Intro",
    "type": "talk_topic",
    "dynamic_line": {
      "concatenate": [
        "&Before you ask, we're full",
        {
          "follower_present": "NC_STARTINGNPC_LIAM",
          "yes": "?\"\n\nLiam looks around at the crowded room, speaking in a soft voice for just your ears.  \"Can't honestly believe people are trying to crowd in here.  They're all so scared, this is a death trap.",
          "no": "?"
        } 
      ]
    },
    "speaker_effect": { "effect": { "npc_add_var": "asked_about_exodii_warehouse_mission", "value": "yes" } },
    "responses": [
      {
        "//": "The other responses are stored in common_talk.json in the exodii folder, to avoid repetition.",
        "text": "Oh well.  Thanks anyway.  I'll go ask around.",
        "topic": "TALK_DONE",
        "condition": { "npc_has_var": "exodii_mission_wh_correct", "value": "no" },
        "effect": [
          { "math": [ "exodii_mission_wh_chance", "-=", "1" ] },
          { "math": [ "exodii_mission_wh_chance", "=", "max( exodii_mission_wh_chance, 1 )" ] }
        ]
      }
    ]
  }
```

### CODE REVIEW

File: data/json/npcs/refugee_center/surface_staff/Smokes/free_merchant_shopkeep_talk.json [hunk 1]
```diff
@@ -204,7 +214,16 @@
...
+    "dynamic_line": {
+      "concatenate": [
+      "I don't know, why don't you ask them?  I haven't pried, but where else do they have to go?  You've seen how bad it is out there.  We don't have the space for them, but it won't kill us to give them some shelter as long as they aren't too much of a drain on resources",
+      {
+        "follower_present": "NC_STARTINGNPC_LIAM",
+        "yes": ".\"\n\n\"Don't you guys have, like, scavengers and things?\"  Liam sounds like he's trying not to get irritated with Smokes.  \"Can't these guys help out in some way in exchange for food?  It seems insane to just keep them sitting in your entry way.\"\n\nSmokes gives him a long-suferring look.  \"Like I said, ask them.  We can't take them in, though.  It's not just them.  We've had to turn away dozens of people, if we started making exceptions we'd be dead meat.  I don't like this any more than you do.",
```
Comment (Karol1223) [inline data/json/npcs/refugee_center/surface_staff/Smokes/free_merchant_shopkeep_talk.json:222]:
```suggestion
        "yes": ".\"\n\n\"Don't you guys have, like, scavengers and things?\"  Liam sounds like he's trying not to get irritated with Smokes.  \"Can't these guys help out in some way in exchange for food?  It seems insane to just keep them sitting in your entry way.\"\n\nSmokes gives him a long-suffering look.  \"Like I said, ask them.  We can't take them in, though.  It's not just them.  We've had to turn away dozens of people, if we started making exceptions we'd be dead meat.  I don't like this any more than you do.",
```

**Review by Karol1223 — CHANGES_REQUESTED** [review]

File: src/condition.cpp [hunk 2]
```diff
@@ -1427,6 +1427,29 @@ conditional_t::func f_npc_train_spells( bool is_npc )
     };
 }
 
+conditional_t::func f_follower_present(const JsonObject &jo, std::string_view member)
+{
+    dbl_or_var empty;
+    const std::string &var_name = jo.get_string(std::string(member));
+    return [var_name](dialogue const &d)
+    {
+        shared_ptr_fast<npc> npc_to_get = overmap_buffer.find_npc_by_unique_id(std::string(var_name));
```
Comment (RenechCDDA) [inline src/condition.cpp:1436]: Unique_ID is only used to track the refugee center guards, you'd want to compare their class IDs like this

```suggestion
npc *npc_to_get;
for( npc &guy : all_npcs() ) {
     if ( guy.myclass == var_name ) {
     npc_to_get = guy;
     break;
     }
```

Comment (katemonster33) [reply src/condition.cpp]: thanks, done

File: data/json/npcs/refugee_center/surface_staff/Smokes/free_merchant_shopkeep_talk.json (hunk 1 above)
Comment (katemonster33) [reply data/json/npcs/refugee_center/surface_staff/Smokes/free_merchant_shopkeep_talk.json]: the json change wasn't meant to be part of the PR, i've removed it. use the snippet in the PR description to test this yourself

**Comment by MNG-cataclysm** [comment]
Come to think of it, this would be a useful way to check if Barry is following you for #73264.

File: src/condition.cpp [hunk 3]
```diff
@@ -1427,6 +1427,34 @@ conditional_t::func f_npc_train_spells( bool is_npc )
     };
 }
 
+conditional_t::func f_follower_present( const JsonObject &jo, std::string_view member )
+{
+    dbl_or_var empty;
```
Comment (anothersimulacrum) [inline src/condition.cpp:1432]: Unused?

File: data/json/npcs/exodii/common_talk.json [hunk 4]
```diff
@@ -46,15 +46,41 @@
     "type": "talk_topic",
     "id": "TALK_Directions_exodii_true",
     "//": "For exodii 'find warehouse' mission, used by NPCs giving directions.",
-    "dynamic_line": "&You listen to the directions, and jot them down in your map.",
+    "dynamic_line": {
+      "concatenate": [
+        "&You listen to the directions, and jot them down in your map",
+        {
+          "and": [ { "math": [ "rng(1, 3)", ">", "1" ] }, { "follower_present": "NC_STARTINGNPC_LIAM" } ],
```
Comment (I-am-Erk) [inline data/json/npcs/exodii/common_talk.json:53]:
```suggestion
          "and": [ { "one_in_chance": 2 }, { "follower_present": "NC_STARTINGNPC_LIAM" } ],
```

Comment (katemonster33) [reply data/json/npcs/exodii/common_talk.json]: i see an error now when loading the game related to this line. i dont really understand enough about eoc to know whats wrong though

File: src/condition.cpp (hunk 3 above)
Comment (katemonster33) [reply src/condition.cpp]: indeed. done.

**Comment by I-am-Erk** [comment]
I am not quite sure if this counts as merging my own PR or not... but who is gonna stop me? HAHAHAHAHA.

**Comment by I-am-Erk** [comment]
Nobody. See? Nobody could. They'd have had to be very quick.

**Comment by I-am-Erk** [comment]
Silliness aside, thanks *so much* for this, Kate.

**Pull request merged**